core/src/main/java/org/lflang/analyses/dag/Dag.java

```
@@ -0,0 +1,319 @@
+        System.exit(1);
+      }
+
+      // Add debug message, if any.
+      label += node.getDotDebugMsg().equals("") ? "" : "\n" + node.getDotDebugMsg();
+      // Add fillcolor and style
+      label += "\", fillcolor=\"" + node.getColor() + "\", style=\"filled\"";
+
+      code += i + "[" + label + "]";
+      dot.pr(code);
+    }
+
+    // Align auxiliary nodes.
+    dot.pr("{");
+    dot.indent();
+    dot.pr("rank = same;");
+    for (Integer i : auxiliaryNodes) {
+      dot.pr(i + "; ");
+    }
+    dot.unindent();
+    dot.pr("}");
+
+    // Add edges
+    for (DagNode source : this.dagEdges.keySet()) {
+      HashMap<DagNode, DagEdge> inner = this.dagEdges.get(source);
+      if (inner != null) {
+        for (DagNode sink : inner.keySet()) {
+          int sourceIdx = dagNodes.indexOf(source);
+          int sinkIdx = dagNodes.indexOf(sink);
+          dot.pr(sourceIdx + " -> " + sinkIdx);
+        }
+      }
+    }
+
+    dot.unindent();
+    dot.pr("}");
+
+    return this.dot;
+  }
+
+  public void generateDotFile(Path filepath) {
+    try {
+      CodeBuilder dot = generateDot();
+      String filename = filepath.toString();
+      dot.writeToFile(filename);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Parses the dot file, reads the edges and updates the DAG. We assume that the edges are
+   * specified as: <SrcNodeId> -> <SinkNodeId>.
+   *
+   * <p>Furthermore, we assume that the new DAG (contained in the dotFile) will not have unnecessary
+   * edges, since they are removed by the shceduler.
+   *
+   * @param dotFilename
+   * @return
+   */
+  public boolean updateDag(String dotFileName) throws IOException {
+    FileReader fileReader;
+    BufferedReader bufferedReader;
+    // Read the file
+    try {
+      fileReader = new FileReader(dotFileName);
+      // Buffer the input stream from the file
+      bufferedReader = new BufferedReader(fileReader);
+    } catch (IOException e) {
```

These should be put in a try-with-resources (TWR). The close method should be in a finally block, which isn't necessary with a TWR.

```java
try (BufferedReader bufferedReader = Files.newBufferedReader(Paths.get(dotFileName))) {
 // your code here
} catch (IOException e) {
  // your handler here
}
// no need to close explicitly
```

The exception handler doesn't look useful either. I think you should let the IOException propagate to the caller and let the caller handle it in a more sensible way. You can just omit the catch block of your TWR to do that. The boolean return value of this method would not be needed.